clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -587,31 +633,57 @@ protected void onJoinPrepare(int generation, String memberId) {
 
         // execute the user's callback before rebalance
-        ConsumerRebalanceListener listener = subscriptions.rebalanceListener();
-
+        final Set<TopicPartition> revokedPartitions;
+        final AtomicReference<Exception> firstException = new AtomicReference<>(null);
         switch (protocol) {
             case EAGER:
-                Set<TopicPartition> revokedPartitions = new HashSet<>(subscriptions.assignedPartitions());
-                log.info("Revoking previously assigned partitions {}", revokedPartitions);
-                try {
-                    listener.onPartitionsRevoked(revokedPartitions);
-                } catch (WakeupException | InterruptException e) {
-                    throw e;
-                } catch (Exception e) {
-                    log.error("User provided listener {} failed on partition revocation", listener.getClass().getName(), e);
-                }
-
-                // also clear the assigned partitions since all have been revoked
+                // revoke all partitions
+                revokedPartitions = new HashSet<>(subscriptions.assignedPartitions());
                 subscriptions.assignFromSubscribed(Collections.emptySet());
 
+                firstException.compareAndSet(null, maybeRevokePartitions(revokedPartitions));
                 break;
 
             case COOPERATIVE:
+                // only revoke those partitions that are not in the subscription any more.
+                Set<TopicPartition> ownedPartitions = new HashSet<>(subscriptions.assignedPartitions());
+                revokedPartitions = ownedPartitions.stream()
+                    .filter(tp -> !subscriptions.subscription().contains(tp.topic()))
+                    .collect(Collectors.toSet());
+
+                if (!revokedPartitions.isEmpty()) {
+                    ownedPartitions.removeAll(revokedPartitions);
+                    subscriptions.assignFromSubscribed(ownedPartitions);
+
+                    firstException.compareAndSet(null, maybeRevokePartitions(revokedPartitions));
+                }
+
                 break;
         }
 
         isLeader = false;
         subscriptions.resetGroupSubscription();
+
+        if (firstException.get() != null) {
+            throw new KafkaException("User rebalance callback throws an error", firstException.get());
+        }
+    }
+
+    @Override
+    public void resetGeneration(String errorMessage) {
+        if (subscriptions.partitionsAutoAssigned()) {
+            // lost all partitions if it is based on subscriptions
+            Set<TopicPartition> lostPartitions = new HashSet<>(subscriptions.assignedPartitions());
+            subscriptions.assignFromSubscribed(Collections.emptySet());
+
+            Exception e = maybeLostPartitions(errorMessage, lostPartitions);
+
+            if (e != null) {
+                throw new KafkaException("User rebalance callback throws an error", e);
```

Comment:
Hmm.. Do we know how this is getting propagated in all cases? Some of the responses are handled using a `RequestCompletionHandler`, but NetworkClient currently eats any exception raised from this callback.

Reply:
Good point, will make a pass through all the call traces.

Reply:
Just to confirm: for AbstractCoordinator we have a specific `CoordinatorResponseHandler` which captures the `RuntimeException` (and KafkaException extends it) and raise to ConsumerCoordinator. So I think that path is covered.